interfaces/builtin/serial_port.go

```diff
@@ -208,6 +210,64 @@ func (iface *serialPortInterface) AutoConnect(*snap.PlugInfo, *snap.SlotInfo) bo
 	return true
 }
 
+func (iface *serialPortInterface) HotplugDeviceDetected(di *hotplug.HotplugDeviceInfo) (*hotplug.ProposedSlot, error) {
+	bus, _ := di.Attribute("ID_BUS")
+	if di.Subsystem() != "tty" || bus != "usb" || !serialDeviceNodePattern.MatchString(di.DeviceName()) {
+		return nil, nil
+	}
+
+	slot := hotplug.ProposedSlot{
+		Attrs: map[string]interface{}{
+			"path": di.DeviceName(),
+		},
+	}
+	if vendor, ok := di.Attribute("ID_VENDOR_ID"); ok {
+		slot.Attrs["usb-vendor"] = vendor
+	}
+	if product, ok := di.Attribute("ID_MODEL_ID"); ok {
+		slot.Attrs["usb-product"] = product
+	}
+	return &slot, nil
+}
+
+func compareSlotAndDeviceAttribute(slotAttributeValue int64, di *hotplug.HotplugDeviceInfo, devinfoAttribute string) bool {
+	var attr string
+	var ok bool
+	if attr, ok = di.Attribute(devinfoAttribute); !ok {
+		return false
+	}
+	val, err := strconv.ParseInt(attr, 16, 64)
+	return err == nil && val == slotAttributeValue
+}
+
+func (iface *serialPortInterface) HandledByGadget(di *hotplug.HotplugDeviceInfo, slot *snap.SlotInfo) bool {
+	// if the slot has vendor, product and interface number set, check if they match
+	var usbVendor, usbProduct, usbInterfaceNumber int64
+	if err := slot.Attr("usb-vendor", &usbVendor); err == nil {
```

Comment:
should we use hasUsbAttrs here to be consistent with the rest of the code?

Reply:
I tried it and I'm not sure if it improves anything, I understand in some other places of this interface it helps simplify the logic a bit, but here I need to try retrive all 3 attributes anyway the way I do right now.. So I'm not sure it helps here.